### ISSUE TITLE
Exported model with dropout incorrectly applies dropout during eval

### ISSUE THREAD

**Issue by andrewor14** [pytorch/pytorch#103681]
Normally calling `model.eval()` on a model with dropout changes the behavior of the dropout ops into a noop. However, this is not true if model is already exported, since the aten dropout pattern remains in the graph. Using the resulting model for inference led to incorrect computation, since we continue to apply dropout to the inputs during eval, which is unexpected. This affects QAT for the PT2 flow, where all transformations are done on an exported model.

Simple repro:
```
import torch
import torch._dynamo

class MyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.dropout = torch.nn.Dropout(p=0.5)
    def forward(self, x):
        return self.dropout(x)

model = MyModel()
model, _ = torch._dynamo.export(model, torch.randn(1), aten_graph=True)
model.eval()
print(model)
```

This prints:
```
def forward(self, x):
    arg0, = fx_pytree.tree_flatten_spec(([x], {}), self._in_spec)
    empty_like_default = torch.ops.aten.empty_like.default(arg0, memory_format = torch.contiguous_format)
    bernoulli__float = torch.ops.aten.bernoulli_.float(empty_like_default);  empty_like_default = None
    div__scalar = torch.ops.aten.div_.Scalar(bernoulli__float, 0.5);  bernoulli__float = None
    mul_tensor = torch.ops.aten.mul.Tensor(arg0, div__scalar);  arg0 = div__scalar = None
    return pytree.tree_unflatten([mul_tensor], self._out_spec)
```

**Comment by ezyang** [comment]
The easy fix is to make eval() error out on exported modules, this is just not supported and cannot be easily done so.

**Comment by ezyang** [comment]
Yeah, this workflow won't work unless you somehow use PT2 export to export a model that can handle BOTH training and eval. Which can only really be done with torch.cond.

**Comment by kimishpatel** [comment]
> The easy fix is to make eval() error out on exported modules, this is just not supported and cannot be easily done so.

@ezyang  Does this mean that params on the exported model will still have `require_grad=True`?

> Yeah, this workflow won't work unless you somehow use PT2 export to export a model that can handle BOTH training and eval. Which can only really be done with torch.cond.

I dont follow. Can you give an example of "export a model that can handle BOTH training and eval. Which can only really be done with torch.cond."

**Comment by ezyang** [comment]
So let's imagine a trivial model that just calls dropout. Upon export, you naively would end up with a graph like:

```
def graph(x: Tensor):
    y = dropout(x, 0.5, training=False)
    return y
```

or True, depending on if you had exported eval/train. But notice that whether or not you are training or not is burned into the graph. Actually, if you export decomps, there's a decent chance the dropout would have gotten eliminated entirely.

So you actually want something like:

```
def graph(x: Tensor, training: bool):
  y = torch.cond(training,
    lambda: dropout(x, 0.5, training=True),
    lambda: x.detach()
  )
  return y
```

This is the sense in which you need torch.cond.

**Comment by kimishpatel** [comment]
> Does this mean that params on the exported model will still have `require_grad=True`?
> 
> >

Ok the example clarified it. What about this question regarding `requires_grad` on params?

**Comment by jerryzh168** [comment]
> So let's imagine a trivial model that just calls dropout. Upon export, you naively would end up with a graph like:
> 
> ```
> def graph(x: Tensor):
>     y = dropout(x, 0.5, training=False)
>     return y
> ```
> 
> or True, depending on if you had exported eval/train. But notice that whether or not you are training or not is burned into the graph. Actually, if you export decomps, there's a decent chance the dropout would have gotten eliminated entirely.
> 
> So you actually want something like:
> 
> ```
> def graph(x: Tensor, training: bool):
>   y = torch.cond(training,
>     lambda: dropout(x, 0.5, training=True),
>     lambda: x.detach()
>   )
>   return y
> ```
> 
> This is the sense in which you need torch.cond.

@ezyang I think we should support this in order to officially support training on exported model.

**Comment by andrewor14** [comment]
> So let's imagine a trivial model that just calls dropout. Upon export, you naively would end up with a graph like:
> 
> ```
> def graph(x: Tensor):
>     y = dropout(x, 0.5, training=False)
>     return y
> ```
> 
> or True, depending on if you had exported eval/train. But notice that whether or not you are training or not is burned into the graph. Actually, if you export decomps, there's a decent chance the dropout would have gotten eliminated entirely.
> 
> So you actually want something like:
> 
> ```
> def graph(x: Tensor, training: bool):
>   y = torch.cond(training,
>     lambda: dropout(x, 0.5, training=True),
>     lambda: x.detach()
>   )
>   return y
> ```
> 
> This is the sense in which you need torch.cond.

If I understand correctly, you're saying this is the only way for us to switch between `model.train()` and `model.eval()` (in both directions) on an exported model for dropout to behave correctly. That makes sense to me. We could rewrite dropout with this `torch.cond` for QAT as a workaround, but this seems like a more general an issue outside of quantization. Should we do this rewriting as part of export?

**Comment by eellison** [comment]
There are other operators that have implementations between inference and training, such as batchnorm. batchnorm also does mutation in training, but is functional in inference. I don't know if it's feasible to just torch.cond around it. 

I think it makes more sense to export the model twice, once for inference and once for training, if that's possible.

**Comment by kimishpatel** [comment]
> I think it makes more sense to export the model twice, once for inference and once for training, if that's possible

I dont know if "reexport" already exported model is well supported, or what that even means. Although I can understand the need for "reexport", it would probably be better served either by a different API or with some options args in export, e.g. via export_config(export_for_training=True). cc: @avikchaudhuri  and Yanan (cant find his gh handle)

**Comment by kimishpatel** [comment]
> Not calling export again on an already exported model

Well, how would that work? You want to export trained model for inference, not original model.

**Comment by kimishpatel** [comment]
> "Reexporting" wouldn't actually help in this case because we already end up with the aten dropout pattern for train/eval after the first export, so it's too late to change it.

I meant, if you could make `train_m.eval()` work but it required you to re-export the model.

**Comment by andrewor14** [comment]
> > Not calling export again on an already exported model
> 
> Well, how would that work? You want to export trained model for inference, not original model.

Yes, it wouldn't work. See my comment above.

> I meant, if you could make train_m.eval() work but it required you to re-export the model.

Right, this would require changing export to detect whether the graph is in eval mode, and do the pattern replacement accordingly there. We already do this in QAT [today](https://github.com/pytorch/pytorch/blob/194fe1d12f9860734cc28ed21bdabda2fbb06336/torch/ao/quantization/_pt2e/utils.py#L242-L287) as a workaround. The question is whether we can push this to the export call itself.

**Comment by vadimkantorov** [comment]
IMO also simply disabling `.eval()` on exported models is better and clear than introducing a not very idiomatic `.move_model_to_eval()`
